lib/stm32/f7/desig.c

```diff
@@ -0,0 +1,71 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2019 Matthew Lai <m@matthewlai.ca>
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <libopencm3/stm32/dbgmcu.h>
+#include <libopencm3/stm32/desig.h>
+
+uint16_t desig_get_flash_size(void)
+{
+  uint32_t device_id = DBGMCU_IDCODE & DBGMCU_IDCODE_DEV_ID_MASK;
+  uint32_t* flash_size_base = 0;
+  switch (device_id) {
+    case 0x0449:
+      flash_size_base = (uint32_t*)DESIG_FLASH_SIZE_BASE_449;
+      break;
+    case 0x0451:
+      flash_size_base = (uint32_t*)DESIG_FLASH_SIZE_BASE_451;
+      break;
+    case 0x0452:
+      flash_size_base = (uint32_t*)DESIG_FLASH_SIZE_BASE_452;
+      break;
+  }
+
+  if (!flash_size_base) {
+    /* We don't know the address for this device. Hang here to help debugging. */
+    for(;;);
```

Comment:
perhaps nicer to return 0?  we have a cm3_assert you could use too, cm3_assert_not_reached iirc, or similar.

Reply:
I thought about returning 0, but thought it would be nicer to fail loudly rather than silently. What do you think?

cm3_assert_not_reached() is nice. I'll change it tonight when I'm back home.

Reply:
Done.